# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def add_two(x):
    return x + 2

print(add_two)
add_two

<function add_two at 0x1060e3920>


<function __main__.add_two(x)>

In [3]:
# lambda param: one-line function
lambda x: x + 2

<function __main__.<lambda>(x)>

In [3]:
print(lambda x: x + 2)  # This is an annoymous functon.  See 

<function <lambda> at 0x00000251CCDF8790>


In [4]:
add_two(10)

12

In [5]:
# IIFE - Immediately Invoked Function Expression
# Execute a lambda function immediately after defining the function
# Syntax: (lambda param: expression)(arg)

(lambda x: x+2)(10)

12

In [6]:
num = 25

num_plus_two = (lambda x: x + 2)(num)

print(num_plus_two)

27


#### Saving to a Variable

In [4]:
#Saving lamda to a variable is not often done:

named_lambda = lambda y: y**2

print(named_lambda)

<function <lambda> at 0x106148720>


In [8]:
named_lambda(5)

25

#### Multiple Inputs

In [5]:
def multiply(x, y, z):
    return x * y * z

multiply(10, 20, 2)

400

In [6]:
multiply(5, 6, 7)

210

In [7]:
# Syntax: lambda param1, param2, etc: expression

lambda x, y, z: x * y * z

<function __main__.<lambda>(x, y, z)>

In [9]:
# As an IIFE - (lambda param1, param2, etc: expression)(arg1, arg2, etc.)
(lambda x, y, z: x * y * z)(2, 3, 4)

24

In [11]:
def accept_args(*args):
    return args

accept_args(1, 2, 3, 4)

(1, 2, 3, 4)

In [12]:
(lambda *args: args)('a', 'b', 'c')

('a', 'b', 'c')

In [14]:
def accept_kwargs(**kwargs):
    return kwargs

accept_kwargs(name='Brian', age=99, city='Chicago')

{'name': 'Brian', 'age': 99, 'city': 'Chicago'}

In [15]:
(lambda **kwargs: kwargs)(name='Sarah', age=55, city='Canyon City')

{'name': 'Sarah', 'age': 55, 'city': 'Canyon City'}

In [17]:
def get_evens(nums):
    output = []
    for num in nums:
        if num % 2 == 0:
            output.append(num)
    return output

get_evens([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])

[2, 4, 6, 8, 10]

In [18]:
(lambda nums: [num for num in nums if num % 2 == 0])([5, 10, 15, 20, 25, 30, 35, 40, 45, 50])

[10, 20, 30, 40, 50]

#### Passing a Function into a Function

In [19]:
def apply_to_each(a_list, f):
    """
    Print the return value of function f applied to every element in a_list
    a_list -> List of elements to execute function f on
    f -> Function to execute
    """
    for element in a_list:
        f_return = f(element)
        print(f_return)

In [21]:
def get_first_letter(string):
    return string[0]

In [22]:
months = ["jan", "feb", "mar", "apr"]

In [23]:
apply_to_each(months, get_first_letter)

j
f
m
a


In [35]:
list_of_seasons = ['Winter', 'Spring', 'Summer', 'Fall']

In [36]:
apply_to_each(list_of_seasons, get_first_letter)

W
S
S
F


In [37]:
def reverse_string(string):
    return string[::-1]

In [38]:
apply_to_each(list_of_seasons, reverse_string)

retniW
gnirpS
remmuS
llaF


#### Passing a Lambda into a Function

In [39]:
# Call the apply_to_each function, pass in a lambda function as f argument

apply_to_each(list_of_seasons, lambda season: season.upper())

WINTER
SPRING
SUMMER
FALL


In [40]:
apply_to_each([5, 10, 15, 20, 25, 30], lambda num: num**2 if num % 2 == 0 else num * 2)

10
100
30
400
50
900


In [42]:
def double_or_square(num):
    if num % 2 == 0:
        return num ** 2
    else:
        return num * 2

apply_to_each([5, 10, 15, 20, 25, 30], double_or_square)

10
100
30
400
50
900


#### Returning a Function from a Function

In [47]:
def my_outer_function(name):
    print(f"Hello {name}")
    
    # Define a function - will be scoped to my_outer_function
    def my_inner_function():
        city = 'Chicago'
        print(f"This is the inner function {name} from {city}")
        
    # Execute the inner function
    my_inner_function()
    
#     print(f"Goodbye {name} from {city}") # NameError: name 'city' is not defined
    
    
print(my_outer_function)
my_outer_function

<function my_outer_function at 0x00000251CCDF89D0>


<function __main__.my_outer_function(name)>

In [48]:
my_outer_function('Brian')

Hello Brian
This is the inner function Brian from Chicago


In [50]:
# my_inner_function()

In [51]:
my_outer_return = my_outer_function('Sarah')

Hello Sarah
This is the inner function Sarah from Chicago


In [52]:
print(my_outer_return)

None


In [53]:
def my_outer_function2(name):
    print(f"Hello {name}")
    
    # Define a function - will be scoped to my_outer_function
    def my_inner_function():
        city = 'Chicago'
        print(f"This is the inner function {name} from {city}")
        
    # return the inner function
    return my_inner_function


print(my_outer_function2)
my_outer_function2

<function my_outer_function2 at 0x00000251CF086950>


<function __main__.my_outer_function2(name)>

In [55]:
my_outer_return2 = my_outer_function2('Brian')
print(my_outer_return2)

Hello Brian
<function my_outer_function2.<locals>.my_inner_function at 0x00000251CF086D40>


In [56]:
my_outer_return2()

This is the inner function Brian from Chicago


In [57]:
def create_multiplier(num):
    def multiply(x):
        return num * x
    return multiply

five_times = create_multiplier(5)
print(five_times)

<function create_multiplier.<locals>.multiply at 0x00000251CF086440>


In [59]:
five_times(10)

50

In [60]:
five_times(3)

15

In [62]:
three_times = create_multiplier(3)
print(three_times)

<function create_multiplier.<locals>.multiply at 0x00000251CCD47EB0>


In [63]:
three_times(10)

30

In [64]:
three_times(3)

9

In [65]:
equals_line = create_multiplier('=')

equals_line(50)

'=================================================='

#### Returning a Lambda from a Function

In [67]:
def new_multiplier(num):
    return lambda x: num * x


four_times = new_multiplier(4)
print(four_times)
four_times

<function new_multiplier.<locals>.<lambda> at 0x00000251CF0879A0>


<function __main__.new_multiplier.<locals>.<lambda>(x)>

In [69]:
print(four_times(10))
print(four_times(5))

40
20


#### If Statements within Lambdas

In [70]:
def double_or_square(num):
    if num % 2 == 0:
        return num ** 2
    else:
        return num * 2

print(double_or_square(10))
print(double_or_square(5))
print(double_or_square(12))
print(double_or_square(7))

100
10
144
14


In [71]:
# Ternary: value_if_true if condition else value_if_false
def double_or_square(num):
    return num ** 2 if num % 2 == 0 else num * 2

print(double_or_square(10))
print(double_or_square(5))
print(double_or_square(12))
print(double_or_square(7))

100
10
144
14


In [72]:
# As an IIFE
(lambda num: num ** 2 if num % 2 == 0 else num * 2)(6)

36

In [73]:
(lambda num: num ** 2 if num % 2 == 0 else num * 2)(9)

18

#### In-Class Exercise #1 <br>
<p>Write a lambda function that takes in three arguments <code>a</code>,<code>b</code>,<code>c</code> and returns the product of <code>a</code> and the difference of <code>b</code> and <code>c</code></p>

In [25]:
# Example 1 
# a = 10, b = 5, c = 2
# Output: 30

# Example 2
# a = 20, b = 10, c = 5
# Output: 100

a = 10 
b = 5 
c = 2

result = (lambda a, b, c: a * (b - c))(20, 10, 5)
print(result)



100


100

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

In [85]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [89]:
# Function that we want to apply to each element in an iterable
def squared_minus_two(num):
    return num**2 - 2

nums = [3, 6, 3, 10, 7, 8, 4, 5]

# Create an empty list
new_nums = []
# Loop over the list
for num in nums:
    # apply the function to each element
    new_num = squared_minus_two(num)
    # Append the new num to new list
    new_nums.append(new_num)
    
print(new_nums)

[7, 34, 7, 98, 47, 62, 14, 23]


In [91]:
# map(function, *iterables -> List, Tuple, String, Set, Dictionary, etc)

list(map(squared_minus_two, nums))

[7, 34, 7, 98, 47, 62, 14, 23]

In [92]:
my_map_obj = map(squared_minus_two, nums)
print(my_map_obj)

In [93]:
help(next)

Help on built-in function next in module builtins:

next(...)
    next(iterator[, default])
    
    Return the next item from the iterator. If default is given and the iterator
    is exhausted, it is returned instead of raising StopIteration.



In [102]:
next(my_map_obj)

StopIteration: 

In [103]:
my_map_obj = map(squared_minus_two, nums)

for num in my_map_obj:
    print(num)

7
34
7
98
47
62
14
23


In [104]:
for num in my_map_obj:
    print(num)

In [107]:
x_values = [1, 2, 3, 4, 5]
y_values = [10, 20, 30, 40, 50]

def multiply(x, y):
    print(f"x={x}, y={y}")
    return x*y

list(map(multiply, x_values, y_values))

x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50


[10, 40, 90, 160, 250]

In [115]:
x_values2 = [1, 2, 3, 4, 5]
y_values2 = [10, 20, 30, 40, 50, 60, 70, 80, 90]

list(map(multiply, x_values2, y_values2))

x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50


[10, 40, 90, 160, 250]

In [117]:
nums_mult = []

shorter = len(x_values2) if len(x_values2) < len(y_values2) else len(y_values2)

for i in range(shorter):
    nums_mult.append(multiply(x_values2[i], y_values2[i]))
    
print(nums_mult)

x=1, y=10
x=2, y=20
x=3, y=30
x=4, y=40
x=5, y=50
[10, 40, 90, 160, 250]


In [121]:
def hello_func(element):
    return element.upper()


my_list = ['a', 'b', 'c', 'd', 'e']

list(map(hello_func, my_list))

['A', 'B', 'C', 'D', 'E']

#### Using Lambda's with Map

In [129]:
letters = ['a', 'b', 'c', 'd', 'e']

upper_letters = list(map(lambda l: l.upper(), letters))
print(upper_letters)

['A', 'B', 'C', 'D', 'E']


In [126]:
nums = [2, 4, 6, 8, 10]

list(map(lambda x: x**3, nums))

[8, 64, 216, 512, 1000]

In [127]:
for val in map(lambda x:x**2, nums):
    print(val)

4
16
36
64
100


In [132]:
list(map(lambda x: x**3,nums))

[8, 64, 216, 512, 1000]

In [133]:
alphabet = 'abcdefghijklmnopqrstuvwxyz'

unicode = list(map(lambda letter: ord(letter), alphabet))
print(unicode)

[97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122]


In [27]:
vowels = ('a', 'e', 'i', 'o', 'u')

upper_vowels = tuple(map(lambda v: v.upper(), vowels))

print(upper_vowels)

('A', 'E', 'I', 'O', 'U')


In [136]:
squares = list(map(lambda x: x**2, range(10,21)))
print(squares)

[100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number if it is even and halve it if it is odd</p>

In [37]:
numbers = range(1,11)

result = list(map(lambda x: x * 2 if x % 2==0 else x/2, numbers ))

print(result)

def double_even (number):
    if number % 2 ==0:
        return number *2
    else:
        return number /2
    



# [0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]


[0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]


[0.5, 4, 1.5, 8, 2.5, 12, 3.5, 16, 4.5, 20]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

In [141]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [33]:
names = ['Bob', 'Billy', 'Alex', 'Barbara', 'Amy', 'Andy', 'Catherine', 'Cameron']

def starts_with_a(name):
    return name[0].upper() == 'A'

a_names_map = list(map(starts_with_a, names))
print(a_names_map)

[False, False, True, False, True, True, False, False]


In [143]:
# Syntax: filter(function_to_apply, iterable -> List,Tuple,Set,Dict,String,etc)
a_names_filter = list(filter(starts_with_a, names))
print(a_names_filter)

['Alex', 'Amy', 'Andy']


In [34]:
nums = [num for num in range(1,11)]

def remainder(num):
    # Will return the remainder of num after dividing by 3
    return num % 3

print(nums)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


In [152]:
nums_map = list(map(remainder, nums))
print(nums_map)

[1, 2, 0, 1, 2, 0, 1, 2, 0, 1]


In [154]:
nums_filter = list(filter(remainder, nums))
print(nums_filter)

[1, 2, 4, 5, 7, 8, 10]


In [156]:
for i in range(-5, 6):
    print(i, bool(i))

-5 True
-4 True
-3 True
-2 True
-1 True
0 False
1 True
2 True
3 True
4 True
5 True


#### Using Lambda's with Filter()

In [157]:
print(names)

['Bob', 'Billy', 'Alex', 'Barbara', 'Amy', 'Andy', 'Catherine', 'Cameron']


In [158]:
b_names = list(filter(lambda name: name[0].upper() == 'B', names))
print(b_names)

['Bob', 'Billy', 'Barbara']


In [159]:
sentence = 'I live at 123 Real Street, City, State 58583. My number is (555) 555-4321'

digits = list(filter(lambda char: char.isdigit(), sentence))
print(digits)

['1', '2', '3', '5', '8', '5', '8', '3', '5', '5', '5', '5', '5', '5', '4', '3', '2', '1']


#### In-Class Exercise #3 <br>
<p>Use the filter function to that take in a list of cities as strings and returns a list of cities with at least two words in the name</p>

In [36]:
cities = ['Chicago', 'Los Angeles', 'New York City', 'Miami', 'Oklahoma City', 'Baton Rouge', 'Boston', 'Denver']
# Output: ['Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge'] 

filter_cities = list(filter(lambda city: len(city.split()) >=2, cities))

print(filter_cities)

['Los Angeles', 'New York City', 'Oklahoma City', 'Baton Rouge']


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

In [172]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


In [176]:
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, iterable[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence
    or iterable, from left to right, so as to reduce the iterable to a single
    value.  For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the iterable in the calculation, and serves as a default when the
    iterable is empty.



#### Syntax

In [177]:
def add_nums(x, y):
    print('='*20)
    print(f"x: {x}")
    print(f"y: {y}")
    print(f"returns: {x + y}")
    return x + y

In [178]:
result_add = reduce(add_nums, [1, 2, 3, 4, 5])
print(result_add)

x: 1
y: 2
returns: 3
x: 3
y: 3
returns: 6
x: 6
y: 4
returns: 10
x: 10
y: 5
returns: 15
15


In [179]:
total = 0
for num in [1, 2, 3, 4, 5]:
    total += num
print(total)

15


In [180]:
def find_max(a, b):
    print('='*20)
    print(f"a: {a}")
    print(f"b: {b}")
    print(f"returns: {a if a > b else b}")
    return a if a > b else b

In [187]:
from random import randint
nums = [randint(1,50) for i in range(10)]
print(nums)

[44, 5, 15, 32, 23, 15, 28, 50, 18, 16]


In [188]:
max_val = reduce(find_max, nums)
print(max_val)

a: 44
b: 5
returns: 44
a: 44
b: 15
returns: 44
a: 44
b: 32
returns: 44
a: 44
b: 23
returns: 44
a: 44
b: 15
returns: 44
a: 44
b: 28
returns: 44
a: 44
b: 50
returns: 50
a: 50
b: 18
returns: 50
a: 50
b: 16
returns: 50
50


In [189]:
max_val = nums[0]
for num in nums:
    if num > max_val:
        max_val = num
print(max_val)

50


#### Using Lambda's with Reduce()

In [191]:
print(nums)

[44, 5, 15, 32, 23, 15, 28, 50, 18, 16]


In [190]:
total = reduce(lambda num1, num2: num1 + num2, nums)
print(total)

246


In [192]:
max_val = reduce(lambda a, b: a if a > b else b, nums)
print(max_val)

50


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [193]:
nums = [2, 4, 6, 8, 10]
# Output: 3840


3840


3840


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [ ]:
def capitalize(name):
    # Takes in a name, upper case the first letter and lower case the rest
    return name[0].upper() + name[1:].lower()

def full_name(first, last):
    # take in a first name, capitalize the first name, take in a last name, capitalize the last name and 
    # then returns the concatenated string of first + space + last
    return capitalize(first) + ' ' + capitalize(last)


def greet_person(first_name, last_name, day_of_week):
    person_full_name = full_name(first_name, last_name)
    print(f"Hello {person_full_name}, it sure is beautiful on this {capitalize(day_of_week)}")
    
greet_person('bRIaN', 'stANtOn', 'tUEsdaY')

In [196]:
import time

def func_one():
    print("Function One Started")
    time.sleep(5)
    print("Function One Ended")
    
    
def func_two():
    print("Function Two Started")
    func_one()
    print("Function Two Ended")
    
    
func_two()

Function Two Started
Function One Started
Function One Ended
Function Two Ended


#### Implementing a Base Case

In [198]:
def add_nums(num):
    # Implement a Base Case
    if num <= 1:
        return num
    else:
        # Add the num to the return of the add_nums function with num minus 1
        return num + add_nums(num - 1)
    
add_nums(10)

55

In [200]:
def add_nums(num):
    # Implement a Base Case
    if num <= 1:
        print(f"add_nums({num}) = {num}")
        return num
    else:
        # Add the num to the return of the add_nums function with num minus 1
        print(f"add_nums({num}) = {num} + add_nums({num - 1})")
        return num + add_nums(num - 1)
    
add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

In [202]:

def is_even(number):
    if number == 1:
        return False
    elif number == 0:
        return True
    else:
        return is_even(number - 2)
    
is_even(21)

False

In [204]:
# Can also be used as a way of looping
def loop(a_list):
    if a_list:
        print(a_list[0])
        loop(a_list[1:])
        
        
loop(['red', 'orange', 'yellow', 'green', 'blue', 'red'])

red
orange
yellow
green
blue
red


#### Writing a Factorial Function

In [1]:
# 5! = 5 * 4!
# 4! = 4 * 3!
# ...
# 1! = 1

# factorial(x) = x * factorial(x-1)

def factorial(num):
    if num <= 1:
        return 1
    else:
        return num * factorial(num - 1)
    
    
factorial(5)

120

In [ ]:
# NEED SOME SORT OF BASE CASE

# def factorial(num):
#     return num * factorial(num - 1)

# factorial(3)

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [2]:
lengths = map(len, ['Brian', 'Stanton', 'Coding Temple', 'Instructor'])

print(lengths)

In [5]:
for item in lengths:
    print(item)

5
7
13
10


In [6]:
for item in lengths:
    print(item)

In [23]:
lengths = map(len, ['Brian', 'Stanton', 'Coding Temple', 'Instructor'])


In [28]:
next(lengths)

StopIteration: 

In [7]:
help(iter)

Help on built-in function iter in module builtins:

iter(...)
    iter(iterable) -> iterator
    iter(callable, sentinel) -> iterator
    
    Get an iterator from an object.  In the first form, the argument must
    supply its own iterator, or be a sequence.
    In the second form, the callable is called until it returns the sentinel.



In [10]:
help(next)

Help on built-in function next in module builtins:

next(...)
    next(iterator[, default])
    
    Return the next item from the iterator. If default is given and the iterator
    is exhausted, it is returned instead of raising StopIteration.



In [12]:
greetings_list = ['hi', 'hello', 'hola', 'bonjour', 'hallo', 'salutations', 'aloha', 'salud']

list_iter = iter(greetings_list)
print(list_iter)

In [21]:
next(list_iter)

StopIteration: 

In [22]:
for g in greetings_list:
    print(g)

hi
hello
hola
bonjour
hallo
salutations
aloha
salud


#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [45]:
def my_range_iter(start, stop, step=1):
    while start < stop:
        yield start
        start += step
        
my_gen = my_range_iter(0, 10)
print(my_gen)

<generator object my_range_iter at 0x00000278520ACB30>


In [56]:
next(my_gen)

StopIteration: 

In [58]:
for yielded_val in my_range_iter(10, 20):
    print(f"The value that is yielded as we loop: {yielded_val}")

The value that is yielded as we loop: 10
The value that is yielded as we loop: 11
The value that is yielded as we loop: 12
The value that is yielded as we loop: 13
The value that is yielded as we loop: 14
The value that is yielded as we loop: 15
The value that is yielded as we loop: 16
The value that is yielded as we loop: 17
The value that is yielded as we loop: 18
The value that is yielded as we loop: 19


In [59]:
for val in my_range_iter(0, 1000, 50):
    print(val)

0
50
100
150
200
250
300
350
400
450
500
550
600
650
700
750
800
850
900
950


In [92]:
def read_file(filename):
    with open(filename) as large_file:
        for line in large_file:
            yield line
            
            
file = read_file('whiteboard.py')
print(file)

<generator object read_file at 0x00000278520AEDC0>


In [1]:
for line in file:
    print(line, end='')

In [98]:
test = read_file('test.py')

In [99]:
for line in test:
    print(line, end='')

from unittest import TestCase, main

from whiteboard import solution

class MatchTestCase(TestCase):
    def test_example_one(self):
        self.assertEqual(solution("test"), "e")
    def test_example_two(self):
        self.assertEqual(solution("teeter"), "r")
    def test_example_three(self):
        self.assertEqual(solution("trend"),"t" )
    def test_example_four(self):
        self.assertEqual(solution("aabbcc"),None)



if __name__ == '__main__':
    main()

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [2]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York", "DC", "     "]


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name *case insensitive*...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [3]:
authors = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield", "David hassELHOFF", "Gary A.J. Bernstein"]


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]
`

In [4]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angeles",44),("Miami",29)]



### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p> <p>More information on the Fibonacci Sequence <a href="https://en.wikipedia.org/wiki/Fibonacci_sequence" target="_blank">here</a>. Start the sequence with `0,1,1,...`.</p>



### Exercise #5<br>
<p>Create a generator that takes a number argument and yields that number squared down to 0, then call the generator and iterate through and print the result.</p>

In [5]:
# for i in generator(10):
#     print(i)

# 100
# 81
# 64
# 49
# 36
# 25
# 16
# 9
# 4
# 2
# 1
# 0